### PR TITLE
scale function(_get_mean_var) updated for dense array, speedup upto ~4.65x

### PR DESCRIPTION
**Pull request by ashish615**
Hi,
We are submitting PR for speed up of the _get_mean_var function. 
|                 |  Time(sec)|
| -----------| ----- |
| Original | 18.49 |
| Updated | 3.97 |
| Speedup | 4.65743073 |

experiment setup : AWS r7i.24xlarge

```py
import time
import numpy as np

import pandas as pd

import scanpy as sc
from sklearn.cluster import KMeans

import os
import wget

import warnings



warnings.filterwarnings('ignore', 'Expected ')
warnings.simplefilter('ignore')
input_file = "./1M_brain_cells_10X.sparse.h5ad"

if not os.path.exists(input_file):
    print('Downloading import file...')
    wget.download('https://rapids-single-cell-examples.s3.us-east-2.amazonaws.com/1M_brain_cells_10X.sparse.h5ad',input_file)


# marker genes
MITO_GENE_PREFIX = "mt-" # Prefix for mitochondrial genes to regress out
markers = ["Stmn2", "Hes1", "Olig1"] # Marker genes for visualization

# filtering cells
min_genes_per_cell = 200 # Filter out cells with fewer genes than this expressed
max_genes_per_cell = 6000 # Filter out cells with more genes than this expressed

# filtering genes
min_cells_per_gene = 1 # Filter out genes expressed in fewer cells than this
n_top_genes = 4000 # Number of highly variable genes to retain

# PCA
n_components = 50 # Number of principal components to compute

# t-SNE
tsne_n_pcs = 20 # Number of principal components to use for t-SNE

# k-means
k = 35 # Number of clusters for k-means

# Gene ranking

ranking_n_top_genes = 50 # Number of differential genes to compute for each cluster

# Number of parallel jobs
sc._settings.ScanpyConfig.n_jobs = os.cpu_count()

start=time.time()
tr=time.time()
adata = sc.read(input_file)
adata.var_names_make_unique()
adata.shape
print("Total read time : %s" % (time.time()-tr))



tr=time.time()
# To reduce the number of cells:
USE_FIRST_N_CELLS = 1300000
adata = adata[0:USE_FIRST_N_CELLS]
adata.shape

sc.pp.filter_cells(adata, min_genes=min_genes_per_cell)
sc.pp.filter_cells(adata, max_genes=max_genes_per_cell)
sc.pp.filter_genes(adata, min_cells=min_cells_per_gene)
sc.pp.normalize_total(adata, target_sum=1e4)
print("Total filter and normalize time : %s" % (time.time()-tr))


tr=time.time()
sc.pp.log1p(adata)
print("Total log time : %s" % (time.time()-tr))


# Select highly variable genes
sc.pp.highly_variable_genes(adata, n_top_genes=n_top_genes, flavor = "cell_ranger")

# Retain marker gene expression
for marker in markers:
        adata.obs[marker + "_raw"] = adata.X[:, adata.var.index == marker].toarray().ravel()

# Filter matrix to only variable genes
adata = adata[:, adata.var.highly_variable]

ts=time.time()
#Regress out confounding factors (number of counts, mitochondrial gene expression)
mito_genes = adata.var_names.str.startswith(MITO_GENE_PREFIX)
n_counts = np.array(adata.X.sum(axis=1))
adata.obs['percent_mito'] = np.array(np.sum(adata[:, mito_genes].X, axis=1)) / n_counts
adata.obs['n_counts'] = n_counts


sc.pp.regress_out(adata, ['n_counts', 'percent_mito'])
print("Total regress out time : %s" % (time.time()-ts))

#scale

ts=time.time()
sc.pp.scale(adata)
print("Total scale time : %s" % (time.time()-ts))
```
add timer around _get_mean_var call 

https://github.com/scverse/scanpy/blob/706d4ef65e5d65e04b788831e7fd65dbe6b2a61f/scanpy/preprocessing/_scale.py#L167

we can also create _get_mean_var_std function that return std as well so we don't require to compute it in scale function(L168-L169).

### CODE REVIEW

**Comment by Intron7** [comment]
I have some small improvements that I would like to add next week for more precision for larger matrices

File: src/scanpy/preprocessing/_utils.py [hunk 1]
```diff
@@ -33,14 +33,43 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+@numba.njit(cache=True, parallel=True)
+def _compute_mean_var(
+    X: _SupportedArray, axis: Literal[0, 1] = 0, dtype: DTypeLike | None = None
+) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
+    nthr = numba.get_num_threads()
+    axis_i = 1 if axis == 0 else 0
+    s = np.zeros((nthr, X.shape[axis_i]), dtype=dtype)
+    ss = np.zeros((nthr, X.shape[axis_i]), dtype=dtype)
+    mean = np.zeros(X.shape[axis_i], dtype=dtype)
+    # std=np.zeros(X.shape[axis_i],dtype=dtype)
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py]:
```suggestion

```

File: src/scanpy/preprocessing/_utils.py [hunk 2]
```diff
@@ -33,14 +33,43 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+        for r in range(i, n, nthr):
+            for c in range(X.shape[axis_i]):
+                v = X[r, c] if axis == 0 else X[c, r]
+                s[i, c] += v
+                ss[i, c] += v * v
+    for c in numba.prange(X.shape[axis_i]):
+        s0 = s[:, c].sum()
+        mean[c] = s0 / n
+        var[c] = (ss[:, c].sum() - s0 * s0 / n) / (n - 1)
+        # std[c]=np.sqrt(var[c])
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py]:
```suggestion

```

File: src/scanpy/preprocessing/_utils.py [hunk 3]
```diff
@@ -33,14 +33,43 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+def _compute_mean_var(
+    X: _SupportedArray, axis: Literal[0, 1] = 0, dtype: DTypeLike | None = None
+) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
+    nthr = numba.get_num_threads()
+    axis_i = 1 if axis == 0 else 0
+    s = np.zeros((nthr, X.shape[axis_i]), dtype=dtype)
+    ss = np.zeros((nthr, X.shape[axis_i]), dtype=dtype)
+    mean = np.zeros(X.shape[axis_i], dtype=dtype)
+    # std=np.zeros(X.shape[axis_i],dtype=dtype)
+    var = np.zeros(X.shape[axis_i], dtype=dtype)
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py:60]:
```suggestion
    s = np.zeros((nthr, X.shape[axis_i]), dtype=np.float64)
    ss = np.zeros((nthr, X.shape[axis_i]), dtype=np.float64))
    mean = np.zeros(X.shape[axis_i], dtype=np.float64))
    # std=np.zeros(X.shape[axis_i],dtype=dtype)
    var = np.zeros(X.shape[axis_i], dtype=np.float64))
```

File: src/scanpy/preprocessing/_utils.py [hunk 4]
```diff
@@ -33,14 +33,41 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+            mean_sq = axis_mean(elem_mul(X, X), axis=axis, dtype=np.float64)
+            var = mean_sq - mean**2
+        # enforce R convention (unbiased estimator) for variance
+        var *= X.shape[axis] / (X.shape[axis] - 1)
+    return mean, var
+
+
+@numba.njit(cache=True, parallel=True)
+def _compute_mean_var(
+    X: _SupportedArray, axis: Literal[0, 1] = 0, dtype: DTypeLike | None = None
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py]:
```suggestion
    X: _SupportedArray, axis: Literal[0, 1] = 0
```

File: src/scanpy/preprocessing/_utils.py [hunk 5]
```diff
@@ -33,14 +33,41 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
 def _get_mean_var(
     X: _SupportedArray, *, axis: Literal[0, 1] = 0
 ) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
-    if isinstance(X, sparse.spmatrix):
-        mean, var = sparse_mean_variance_axis(X, axis=axis)
+    if isinstance(X, np.ndarray):
+        mean, var = _compute_mean_var(X, axis=axis, dtype=np.float64)
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py]:
```suggestion
        mean, var = _compute_mean_var(X, axis=axis)
```

**Comment by Intron7** [comment]
@ashish615 after doing some benchmarking myself I found out that your solution for `axis=1` is under performing compared to `axis=0` for larger arrays. I think that is because of the memory access pattern you choose. I rewrote the function with that in mind. I'll again make a PR to you, because for some reason you disallow us from making changes to your PR.

**Comment by Intron7** [comment]
https://github.com/IntelLabs/open-omics-scanpy/pull/2 This would fix the performance. This is still not perfect for F-Array's but its a start

**Review by Intron7 — CHANGES_REQUESTED** [review]
Please merge https://github.com/IntelLabs/open-omics-scanpy/pull/2

File: docs/release-notes/1.10.2.md [hunk 6]
```diff
@@ -31,3 +31,4 @@
 * `pp.highly_variable_genes` with `flavor=seurat_v3` now uses a numba kernel {pr}`3017` {smaller}`S Dicks`
 * Speed up {func}`~scanpy.pp.scrublet` {pr}`3044` {smaller}`S Dicks` and {pr}`3056` {smaller}`P Angerer`
 * Speed up clipping of array in {func}`~scanpy.pp.scale` {pr}`3100` {smaller}`P Ashish & S Dicks`
+* Speed up _get_mean_var used in {func}`~scanpy.pp.scale` {pr}`3099` {smaller}`P Ashish & S Dicks`
```
Comment (Intron7) [inline docs/release-notes/1.10.2.md]: move this to `0.10.3.md`

**Review by Intron7 — CHANGES_REQUESTED** [review]

File: docs/release-notes/1.10.2.md (hunk 6 above)
Comment (ashish615) [reply docs/release-notes/1.10.2.md]: ok

Comment (ashish615) [reply docs/release-notes/1.10.2.md]: done

File: src/scanpy/preprocessing/_utils.py [hunk 7]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+        axis_i = 1
+        sums = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
+        sums_squared = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
+        mean = np.zeros(X.shape[axis_i], dtype=np.float64)
+        var = np.zeros(X.shape[axis_i], dtype=np.float64)
+        n = X.shape[axis]
+        for i in numba.prange(n_threads):
+            for r in range(i, n, n_threads):
+                for c in range(X.shape[axis_i]):
+                    value = np.float64(X[r, c])
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py]:
```suggestion
                    value = X[r, c]
```

File: src/scanpy/preprocessing/_utils.py [hunk 8]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+            sum_ = sums[:, c].sum()
+            mean[c] = sum_ / n
+            var[c] = (sums_squared[:, c].sum() - sum_ * sum_ / n) / (n - 1)
+    else:
+        axis_i = 0
+        mean = np.zeros(X.shape[axis_i], dtype=np.float64)
+        var = np.zeros(X.shape[axis_i], dtype=np.float64)
+        for r in numba.prange(X.shape[0]):
+            for c in range(X.shape[1]):
+                value = np.float64(X[r, c])
```
Comment (Intron7) [inline src/scanpy/preprocessing/_utils.py]:
```suggestion
                value = X[r, c]
```

**Review by Intron7 — CHANGES_REQUESTED** [review]
This fixes the issues

File: src/scanpy/preprocessing/_utils.py [hunk 9]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+        sums_squared = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
+        mean = np.zeros(X.shape[axis_i], dtype=np.float64)
+        var = np.zeros(X.shape[axis_i], dtype=np.float64)
+        n = X.shape[axis]
+        for i in numba.prange(n_threads):
+            for r in range(i, n, n_threads):
+                for c in range(X.shape[axis_i]):
+                    value = np.float64(X[r, c])
+                    sums[i, c] += value
+                    sums_squared[i, c] += value * value
```
Comment (ilan-gold) [inline src/scanpy/preprocessing/_utils.py]: Remove casting to match previous behavior.  In `1.11.0` we should release a breaking change that changes the behavior of multiplication to preserve the data type better.

**Review by Intron7 — APPROVED** [review]
Looks good to me

File: src/scanpy/preprocessing/_utils.py [hunk 10]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
-    # enforce R convention (unbiased estimator) for variance
-    var *= X.shape[axis] / (X.shape[axis] - 1)
+        if isinstance(X, sparse.spmatrix):
+            mean, var = sparse_mean_variance_axis(X, axis=axis)
+        else:
+            mean = axis_mean(X, axis=axis, dtype=np.float64)
+            mean_sq = axis_mean(elem_mul(X, X), axis=axis, dtype=np.float64)
+            var = mean_sq - mean**2
+        # enforce R convention (unbiased estimator) for variance
+        var *= X.shape[axis] / (X.shape[axis] - 1)
```
Comment (flying-sheep) [inline src/scanpy/preprocessing/_utils.py:47]: Before your change, this line ran unconditionally, now it only runs for the `not isinstance(X, np.ndarray)` case. Is that intentional? Then you should mention that in `_compute_mean_var`’s docstring.

File: src/scanpy/preprocessing/_utils.py [hunk 11]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+    return mean, var
+
+
+@numba.njit(cache=True, parallel=True)
+def _compute_mean_var(
```
Comment (flying-sheep) [inline src/scanpy/preprocessing/_utils.py:52]: We already have `_get_mean_var`. Maybe rename this to `_get_mean_var_ndarray` or `_get_mean_var_dense`?

File: src/scanpy/preprocessing/_utils.py [hunk 12]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+    X: _SupportedArray, axis: Literal[0, 1] = 0, n_threads=1
```
Comment (flying-sheep) [inline src/scanpy/preprocessing/_utils.py:53]:
```suggestion
    X: _SupportedArray, axis: Literal[0, 1] = 0, n_threads: int = 1
```

File: src/scanpy/preprocessing/_utils.py [hunk 13]
```diff
@@ -33,14 +33,55 @@ def _(X: np.ndarray, *, axis: Literal[0, 1], dtype: DTypeLike) -> np.ndarray:
...
+        mean = np.zeros(X.shape[axis_i], dtype=np.float64)
+        var = np.zeros(X.shape[axis_i], dtype=np.float64)
+        for r in numba.prange(X.shape[0]):
+            for c in range(X.shape[1]):
+                value = X[r, c]
+                mean[r] += value
+                var[r] += value * value
+        for c in numba.prange(X.shape[0]):
+            mean[c] = mean[c] / X.shape[1]
+            var[c] = (var[c] - mean[c] ** 2) / (X.shape[1] - 1)
```
Comment (flying-sheep) [inline src/scanpy/preprocessing/_utils.py:83]: Please don’t duplicate identical lines.

```suggestion
    axis_i = 1 - axis
    mean = np.zeros(X.shape[axis_i], dtype=np.float64)
    var = np.zeros(X.shape[axis_i], dtype=np.float64)
    if axis == 0:
        sums = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
        sums_squared = np.zeros((n_threads, X.shape[axis_i]), dtype=np.float64)
        n = X.shape[axis]
        for i in numba.prange(n_threads):
            for r in range(i, n, n_threads):
                for c in range(X.shape[axis_i]):
                    value = X[r, c]
                    sums[i, c] += value
                    sums_squared[i, c] += value * value
        for c in numba.prange(X.shape[axis_i]):
            sum_ = sums[:, c].sum()
            mean[c] = sum_ / n
            var[c] = (sums_squared[:, c].sum() - sum_ * sum_ / n) / (n - 1)
    else:
        for r in numba.prange(X.shape[0]):
            for c in range(X.shape[1]):
                value = X[r, c]
                mean[r] += value
                var[r] += value * value
        for c in numba.prange(X.shape[0]):
            mean[c] = mean[c] / X.shape[1]
            var[c] = (var[c] - mean[c] ** 2) / (X.shape[1] - 1)
```

**Review by flying-sheep — CHANGES_REQUESTED** [review]
I don’t see the claimed speedup in the benchmarks, what’s missing?

Also is `numba.get_num_threads()` safe? E.g. I think `_get_mean_var` is also called in each dask chunks. Will `numba.get_num_threads()` return a reasonable number in that case?

Otherwise nice! I’m not a huge fan of how unpythonic numba code looks, but I don’t think anything can be done about that.

File: src/scanpy/preprocessing/_utils.py (hunk 11 above)
Comment (Intron7) [reply src/scanpy/preprocessing/_utils.py:52]: I think we can rename the kernel

File: src/scanpy/preprocessing/_utils.py (hunk 13 above)
Comment (Intron7) [reply src/scanpy/preprocessing/_utils.py:83]: I think we can slim this down a bit. The two different loops need to be separate though

**Comment by Intron7** [comment]
The function should also work for 1 thread. numba.get_num_threads() is fine it works well with the sparse arrays. But I have no experience with it inside of dask.

File: src/scanpy/preprocessing/_utils.py (hunk 12 above)
Comment (ilan-gold) [inline src/scanpy/preprocessing/_utils.py:53]: I don't think `_SupportedArray` is the right type annotation here.  This doesn't run directly on `dask.Array`, unless I am misunderstanding something.
